### PR TITLE
Add mld support v2

### PR DESCRIPTION
**Pull request by jonludlam**
On top of #422 - this one is another go at doing mld support.

Looks like quite a lot of tidying has been done since my last attempt, so this one is a much smaller diff! Very nice :-)
Not all of the comments on the previous PR (#377) are still valid, but I think I've addressed those that are.

### CODE REVIEW

**Comment by Leonidas-from-XIV** [comment]
@jonludlam Could you rebase upon `main` now that #422 is merged and add a changelog entry?

**Comment by jonludlam** [comment]
Rebased.

**Review by Leonidas-from-XIV — COMMENTED** [review]
Really nice. I'm happy how much shorter this is than #377, guess all the refactoring and unification of parsers has paid off to make it easier to plug additional parsers.

My only concerns are very minor (naming and code generator) but besides that I think that this looks good and we can release it as 2.3 once merged.

**Comment by tmattio** [comment]
Very excited to see this, thanks for the PR @jonludlam!

How do you see this working with Dune? Do you think mdx should default to supporting `.md` and `.mld` or should it be opt-in? If the latter, it might be good to document it somewhere in the Dune documentation

File: lib/mli_parser.ml [hunk 1]
```diff
@@ -150,11 +150,10 @@ let make_block code_block file_contents =
...
 
-let parse_mli file_contents =
-  (* Find the locations of the code blocks within [file_contents], then slice it up into
-     [Text] and [Block] parts by using the starts and ends of those blocks as
-     boundaries. *)
-  let code_blocks = docstring_code_blocks file_contents in
+(* Given the locations of the code blocks within [file_contents], then slice it up into
+   [Text] and [Block] parts by using the starts and ends of those blocks as
+   boundaries. *)
+let parse_generic code_blocks file_contents =
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply lib/mli_parser.ml:156]: yep that's a better name.

File: test/bin/gen_rule_helpers/gen_rule_helpers.ml [hunk 2]
```diff
@@ -52,20 +53,13 @@ type dir = {
...
-  | [] ->
-      Printf.eprintf "No test file for %s\n" dir_name;
-      Printf.eprintf "There should be one of %s, %s, or %s\n" cwd_test_file_md
-        cwd_test_file_t cwd_test_file_mli;
-      exit 1
   | _ ->
-      Printf.eprintf "More than one test file for %s\n" dir_name;
-      Printf.eprintf "There should be only one of %s, %s, or %s\n"
-        cwd_test_file_md cwd_test_file_t cwd_test_file_mli;
+      Format.eprintf "No test file for %s\n" dir_name;
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply test/bin/gen_rule_helpers/gen_rule_helpers.ml]: Fixed

File: test/bin/gen_rule_helpers/gen_rule_helpers.ml [hunk 3]
```diff
@@ -52,20 +53,13 @@ type dir = {
...
+      Format.eprintf "There should be exactly one of [%a]\n" pp_string_list
+        cwd_test_files;
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply test/bin/gen_rule_helpers/gen_rule_helpers.ml:63]: Done

**Review by Leonidas-from-XIV — APPROVED** [review]
Awesome, thanks!

**Comment by Leonidas-from-XIV** [comment]
@tmattio By default dune only checks for `*.md` files, but when manually passing `.mld` files, it works out-of-the-box (provided a version of MDX that has this PR merged): https://github.com/ocaml/dune/pull/7569

I think it would be useful to extend the default glob, either in an newer version of the `mdx` stanza or a newer version of the dune language. Will take this up, but for people wanting to use it now, it's already usable, and will work as soon as 2.3.0 hits opam-repository: https://github.com/ocaml/opam-repository/pull/23670

**Pull request merged**